### PR TITLE
Inductor: fall back bernoulli on cpu

### PR DESCRIPTION
**Pull request by blzheng**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97002

data type: float32
Input size: torch.Size([64, 4, 128, 128])
single socket (32cores):
```
Before: bernoulli 0.001327775239944458 s      dropout 0.0014216173489888509 s
After:  bernoulli 0.0002424612840016683 s     dropout 0.00039757410685221353 s
```

single core:
```
Before: bernoulli 0.04154032731056213 s      dropout 0.04382548745473226 s
After: bernoulli 0.006143261671066284 s      dropout 0.0065830423831939695 s
```

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -444,16 +444,16 @@ def view_copy_dtype(self, dtype):
...
 
-
-@register_extra_random_decomp([aten.bernoulli.p])
-def bernoulli_p(self, p=0.5, *, generator=None):
-    assert generator is None
-    return torch.rand_like(self, dtype=torch.float32) < p
+    @register_extra_random_decomp([aten.bernoulli.p])
+    def bernoulli_p(self, p=0.5, *, generator=None):
+        assert generator is None
+        return torch.rand_like(self, dtype=torch.float32) < p
```
Comment (jgong5) [inline torch/_inductor/decomposition.py:456]: It is possible both cpu and gpu are available while we only want to do fallback on cpu. Suggest to check the device type of `self` and raise `NotImplemented` to go directly to lowering if the device is cpu.

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -1059,8 +1059,8 @@ def native_dropout(x, p, train):
 @register_lowering(aten.bernoulli_, type_promotion_kind=None)
 def bernoulli_(x, *args):
     assert (
-        config.fallback_random
-    ), "this should be handled in decomps unless config.fallback_random"
+        config.fallback_random or not has_triton()
+    ), "this should be handled in decomps unless config.fallback_random or not has_triton()"
```
Comment (jgong5) [inline torch/_inductor/lowering.py:1063]: ditto: revise the checks.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/decomposition.py (hunk 1 above)
Comment (blzheng) [reply torch/_inductor/decomposition.py]: Thanks for your suggestion. I have updated all checks.

File: torch/_inductor/decomposition.py [hunk 3]
```diff
@@ -447,11 +447,15 @@ def view_copy_dtype(self, dtype):
 
 @register_extra_random_decomp([aten.bernoulli_])
 def bernoulli_(self, p=0.5):
+    if self.device == torch.device("cpu"):
+        return NotImplemented
```
Comment (jgong5) [inline torch/_inductor/decomposition.py:451]:
```suggestion
        raise NotImplemented
```

File: torch/_inductor/decomposition.py [hunk 4]
```diff
@@ -447,11 +447,15 @@ def view_copy_dtype(self, dtype):
...
     return self.copy_(torch.rand_like(self, dtype=torch.float32) < p)
 
 
 @register_extra_random_decomp([aten.bernoulli.p])
 def bernoulli_p(self, p=0.5, *, generator=None):
+    if self.device == torch.device("cpu"):
+        return NotImplemented
```
Comment (jgong5) [inline torch/_inductor/decomposition.py:458]: ditto

File: torch/_inductor/overrides.py [hunk 5]
```diff
@@ -35,21 +35,25 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
             config.fallback_random
             and replacements[func] in replacements_using_triton_random
         ):
-            return replacements[func](*args, **kwargs)
+            if func != torch.nn.functional.dropout or (
+                func == torch.nn.functional.dropout
+                and args[0].device != torch.device("cpu")
+            ):
```
Comment (jgong5) [inline torch/_inductor/overrides.py:41]: How about create a set to hold all the functions to fallback on cpu, e.g., `fallbacks_cpu_random`? Then, we check whether `func` is in it or not here without explicitly checking a specific function here.

File: torch/_inductor/overrides.py [hunk 6]
```diff
@@ -35,21 +35,25 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
...
-def replace_fx(gm: torch.fx.GraphModule):
+def replace_fx(gm: torch.fx.GraphModule, example_inputs):
     # Sometimes patch_functions() misses things already in the graph
     for node in reversed(list(gm.graph.nodes)):
         if node.op == "call_function" and node.target in replacements:
             if (
                 config.fallback_random
-                and replacements[node.target] in replacements_using_triton_random
-            ):
+                or example_inputs[0].device == torch.device("cpu")
```
Comment (jgong5) [inline torch/_inductor/overrides.py:55]: safer to check if all example_inputs are on cpu? similar to `fuse_fx`
```python
def fuse_fx(gm: torch.fx.GraphModule, example_inputs):
    is_cpu = all(
        example_input.device == torch.device("cpu")
        for example_input in example_inputs
        if isinstance(example_input, torch.Tensor)
    )
```

File: torch/_inductor/overrides.py [hunk 7]
```diff
@@ -35,21 +35,25 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
...
+            ) and replacements[node.target] in replacements_using_triton_random:
```
Comment (jgong5) [inline torch/_inductor/overrides.py:56]: Do we need to check if it is "dropout"?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/decomposition.py (hunk 3 above)
Comment (blzheng) [reply torch/_inductor/decomposition.py:451]: Its behavior is not as expected if we replace return with raise. I think we can align it with others in this file.
https://github.com/pytorch/pytorch/blob/ee6b19bd4c4f33590697f45dbd492e77dd9b3e34/torch/_inductor/decomposition.py#L249

File: torch/_inductor/overrides.py (hunk 5 above)
Comment (blzheng) [reply torch/_inductor/overrides.py]: Done in [7aa97a7](https://github.com/pytorch/pytorch/pull/97002/commits/7aa97a7d8a8bcc3dd7cf40eaad52223cd3eebc45)

File: torch/_inductor/overrides.py (hunk 6 above)
Comment (blzheng) [reply torch/_inductor/overrides.py]: Changed in [7aa97a7](https://github.com/pytorch/pytorch/pull/97002/commits/7aa97a7d8a8bcc3dd7cf40eaad52223cd3eebc45)

File: torch/_inductor/overrides.py (hunk 7 above)
Comment (blzheng) [reply torch/_inductor/overrides.py]: For now, yes. We verified that aten::bernoulli with MKL optimizations performs much better than rand_like in inductor, but did not observe that rand_like eager implementation outperforming inductor.

File: torch/_inductor/overrides.py [hunk 8]
```diff
@@ -51,6 +59,12 @@ def replace_fx(gm: torch.fx.GraphModule):
                 and replacements[node.target] in replacements_using_triton_random
             ):
                 continue
+            if node.target in fallback_cpu_random and all(
+                example_input.device == torch.device("cpu")
+                for example_input in example_inputs
+                if isinstance(example_input, torch.Tensor)
```
Comment (jgong5) [inline torch/_inductor/overrides.py:65]: can you lift this cpu device check outside the loop just like `fuse_fx` does? we don't have to compute it multiple times in the loop.

File: torch/_inductor/overrides.py [hunk 9]
```diff
@@ -35,14 +35,22 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
             config.fallback_random
             and replacements[func] in replacements_using_triton_random
         ):
-            return replacements[func](*args, **kwargs)
+            if func not in fallback_cpu_random or (
+                func in fallback_cpu_random
```
Comment (jgong5) [inline torch/_inductor/overrides.py:39]: you don't have to check it again here? also, how about combining this if check with the one at line 34?

File: torch/_inductor/decomposition.py (hunk 3 above)
Comment (jgong5) [reply torch/_inductor/decomposition.py:451]: Oh, makes sense.

**Review by jgong5 — CHANGES_REQUESTED** [review]

**Review by jgong5 — APPROVED** [review]

File: torch/_inductor/decomposition.py [hunk 10]
```diff
@@ -447,11 +447,15 @@ def view_copy_dtype(self, dtype):
 
 @register_extra_random_decomp([aten.bernoulli_])
 def bernoulli_(self, p=0.5):
+    if self.device == torch.device("cpu"):
```
Comment (jansel) [inline torch/_inductor/decomposition.py:450]: I don't believe this will be properly marked as a mutation op, so there could be a correctness issue.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/decomposition.py (hunk 10 above)
Comment (blzheng) [reply torch/_inductor/decomposition.py:450]: Sorry, I did not get your point. As for bernoulli_, there is a ut that is passed when fallback_random=True, and the changes here have the same effect as fallback_random=True. Inplace bernoulli on CPU will be handled by [bernoulli_ in lowering](https://github.com/pytorch/pytorch/pull/97002/files#diff-a1b077971cddfabfa0071c5162265066e867bc07721816d95b9cbe58431c38e3L1059-L1067), which can handle mutation properly.https://github.com/pytorch/pytorch/blob/b507d7d798a17d6297871545db0498f6d1d9af07/test/inductor/test_torchinductor.py#L4531-L4542

**Review by jansel — APPROVED** [review]
